vpr/src/place/noc_place_utils.cpp

```diff
@@ -0,0 +1,643 @@
+        double curr_link_bandwidth = curr_link.get_bandwidth_usage();
+
+        // update the link badnwidth based on the user provided state
+        switch (how_to_update_links) {
+            case link_usage_update_state::increment:
+                curr_link.set_bandwidth_usage(curr_link_bandwidth + traffic_flow_bandwidth);
+                break;
+            case link_usage_update_state::decrement:
+                curr_link.set_bandwidth_usage(curr_link_bandwidth - traffic_flow_bandwidth);
+                break;
+            default:
+                break;
+        }
+
+        // check that the bandwidth never goes to negative
+        VTR_ASSERT(curr_link.get_bandwidth_usage() >= 0.0);
+    }
+
+    return;
+}
+
+void re_route_associated_traffic_flows(ClusterBlockId moved_block_router_id, NocTrafficFlows& noc_traffic_flows_storage, NocStorage& noc_model, NocRouting* noc_flows_router, const vtr::vector_map<ClusterBlockId, t_block_loc>& placed_cluster_block_locations, std::unordered_set<NocTrafficFlowId>& updated_traffic_flows, int& number_of_affected_traffic_flows) {
+    // get all the associated traffic flows for the logical router cluster block
+    const std::vector<NocTrafficFlowId>* assoc_traffic_flows = noc_traffic_flows_storage.get_traffic_flows_associated_to_router_block(moved_block_router_id);
+
+    // now check if there are any associated traffic flows
+    if (assoc_traffic_flows != nullptr) {
+        // There are traffic flows associated to the current router block so process them
+        for (auto& traffic_flow_id : *assoc_traffic_flows) {
+            // first check to see whether we have already re-routed the current traffic flow and only re-route it if we haven't already.
+            if (updated_traffic_flows.find(traffic_flow_id) == updated_traffic_flows.end()) {
+                // now update the current traffic flow by re-routing it based on the new locations of its src and destination routers
+                re_route_traffic_flow(traffic_flow_id, noc_traffic_flows_storage, noc_model, noc_flows_router, placed_cluster_block_locations);
+
+                // now make sure we don't update this traffic flow a second time by adding it to the group of updated traffic flows
+                updated_traffic_flows.insert(traffic_flow_id);
+
+                // update global datastructures to indicate that the current traffic flow was affected due to router cluster blocks being swapped
+                affected_traffic_flows[number_of_affected_traffic_flows] = traffic_flow_id;
+                number_of_affected_traffic_flows++;
+            }
+        }
+    }
+
+    return;
+}
+
+void revert_noc_traffic_flow_routes(const t_pl_blocks_to_be_moved& blocks_affected) {
+    // provides the positions where the affected blocks have moved to
+    auto& place_ctx = g_vpr_ctx.placement();
+    auto& noc_ctx = g_vpr_ctx.mutable_noc();
+
+    NocTrafficFlows* noc_traffic_flows_storage = &noc_ctx.noc_traffic_flows_storage;
```

Comment:
Any reason why you didn't use a reference for this variable too?  NocTrafficFlows&  (and then get rid of the "&" at the start of noc_ctx.noc ...)  It would read a bit more cleanly.

Reply:
Added to an issue and will be updated later